sstables/mp_row_consumer.cc

```diff
@@ -80,9 +81,35 @@ atomic_cell make_counter_cell(api::timestamp_type timestamp, bytes_view value) {
     return ccb.build(timestamp);
 }
 
+// See #6130.
+static data_type freeze_types_in_collections(data_type t) {
+    return ::visit(*t, make_visitor(
+    [] (const map_type_impl& typ) -> data_type {
+        return map_type_impl::get_instance(
+                freeze_types_in_collections(typ.get_keys_type()->freeze()),
+                freeze_types_in_collections(typ.get_values_type()->freeze()),
+                typ.is_multi_cell());
+    },
+    [] (const set_type_impl& typ) -> data_type {
+        return set_type_impl::get_instance(
+                freeze_types_in_collections(typ.get_elements_type()->freeze()),
+                typ.is_multi_cell());
+    },
+    [] (const list_type_impl& typ) -> data_type {
+        return list_type_impl::get_instance(
+                freeze_types_in_collections(typ.get_elements_type()->freeze()),
+                typ.is_multi_cell());
+    },
+    [&] (const abstract_type& typ) -> data_type {
+        return std::move(t);
```

Comment:
Doesn't the lambda have to be mutable for this std::move to not become a copy?

Reply:
The lambda's constness applies to the reference, not the thing being referred to.
Remember that `x& const` is the same as `x&`, and `const x& const` is the same as `const x&`. References are always implicitly const.
Thus lambda's adding of const to the reference to `t` is a no-op.

I checked on a small example to confirm:
```
#include <iostream>

class A {
public:
    A() = default;
    A(const A&) {
        std::cout << "copy cons" << std::endl;
    }
    A(A&&) {
        std::cout << "move cons" << std::endl;
    }
};

int main() {
    A x;
    auto l = [&] () {
        //return x;
        return std::move(x);
    };
    auto y = l();
}
```

it prints "move cons".

Reply:
Makes sense. Thanks for checking.